### ISSUE TITLE
Feature request: Graphical GGUF viewer

### ISSUE THREAD

**Issue by ngxson** [ggml-org/llama.cpp#6715]
# Motivation

With the recent introduction of `eval-callback` example, we now having more tools for debugging when working with llama.cpp. However, one of the tool that I feel missing is the ability to dump everything inside a gguf file into a human-readable (and interactive) interface.

Inspired from `huggingface.js` where users can visualize the KV and list of tensors on huggingface.com, I would like to implement the same thing in llama.cpp. I find this helpful in these situations:
- Debugging `convert.py` script when adding a new architecture
- Debugging tokenizers
- Debugging changes related to gguf (model splits for example)
- Debugging tensors (i.e. display N first elements of a tensor, just like `eval-callback`)
- Debugging control vectors
- ... (maybe other usages in the future)

The reason why I can't use `huggingface.js` is because it's based on browser, which make it tricky when reading a huge local file. It also don't have access to quantized types (same for `gguf-py`).

# Possible Implementation

Ideally, I want the implementation to be a binary named `gguf-viewer` that when run, will open a web page in `localhost:8080`. User can then go to the web page to explore the gguf file. It will have these sections:
- Complete list of KV
- Tokenizer-related info (for example: list all tokens, lookup one token)
- List of all tensors

**Comment by ngxson** [comment]
Yes I tries gguf-py but it does not have access to quantized types

**Comment by bandoti** [comment]
This is something I have been planning on working on so I took the liberty to assign this task to myself.

**Comment by ngxson** [comment]
Yes feel free to take this task. Things changed quite a lot since I created this issue, I feel like it's no longer serves my initial goal (to ease the process of adding new models), but would be nice to have something like @ggerganov suggested above!

**Comment by bandoti** [comment]
I came up with an initial set of high-level features regarding the gguf-viewer program (see below). However, I would like some more ideas regarding the issue-reporting process. I am trying to figure out what should go in with a GGUF viewer, and whether a separate tool should be created with broader scope to capture diagnostic information/issue reporting.

While my intent is not necessarily to discuss implementation details at the moment, I think a good solution for the tool is using Python and TKinter with a custom C extension to expose access to the GGML library. And this also goes for the potential diagnostic tool as Python would be great to: (1) spawn the server process; (2) use the OpenAI APIs; (2) tee the logs (if necessary); (3) load C extensions directly (to access GGUF/GGML libraries). 

1. I would like to interactively explore GGUF files.
2. Meta-Data, tensor info, and the actual tensors should be explorable.
3. More than one GGUF file should be able to be opened at once.
4. Tensors should be visible as both a summary and as visual blocks representing the binary data.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         
5. Tensors should be shown as a high level overview and may be “zoomed into” for more details.
6. The GGUF viewer should be minimal on dependencies and be simply deployable with the llama.cpp suite of programs. It should have access to the GGML/GGUF C APIs.
7. Complete list of tokens should be explorable, and should be visible as both strings and numeric values.
8. Use some sort of heatmap to relate tensor types to visual blocks—colour-coded by category.

**Comment by ngxson** [comment]
Yeah the idea seems good.

> Python and TKinter with a custom C extension to expose access to the GGML library

In fact, when I initially created this issue, the reason why I proposed to have this in cpp was because there is no implementation of quantization outside of cpp at that time.

But things changed a lot since then, `gguf-py` has `quants.py` which allow quantize-dequantize using `numpy`

Going a bit further, I think it's also possible to do this completely on web environment (a bit like https://netron.app/ but having lot more gguf-specific functions):
- Building on top of [huggingface/gguf](https://www.npmjs.com/package/@huggingface/gguf) package which allow access to KV metadata
- Using custom dequantization functions (can be either reimplement from python code, or I can expose these methods via my [wasm](https://github.com/ngxson/wllama) binding)
- Leveraging [FileReader API](https://developer.mozilla.org/en-US/docs/Web/API/FileReader) to read the file chunk by chunk, allow loading even big GGUF

**Comment by bandoti** [comment]
@ngxson Interesting projects—I will keep an eye on them!

I notice in the default install, we are not bundling the gguf-py libraries. Is this something we should bundle with the install? Main reason I ask is because I think it's important to make these diagnostic tools something that can work out-of-the-box on a llama.cpp install for those who are not necessarily interested in pulling in several ML Python dependencies. If it is not something we want to include by default, then naturally a Python extension would make more sense as it can just wrap the ggml library directly.

The "user journey" I imagine is: (1) I have an issue with my model—or I'm curious about a new model; (2) double-click _gguf-viewer.py_ to open a GUI; (3) open a model file—(or select an HF URI to download); (4) explore the model; (5) generate a report and attach to github issue.

I think a lot of people can benefit from this local-first approach, as it reduces the barrier to entry and makes the diagnostic tools more portable in that sense. Even popping open a browser and having a separate server process introduces cognitive load with firewall warnings, and so forth.

That being said, I would like to understand user journeys from the online-first perspective as well—as in on the cloud. I am thinking it is fair to pull meta-data from the models (as we have now), but when it comes to "zooming in" to view the tensor blocks might pose some other issues. Perhaps we can satisfy both needs somehow.

**Comment by bandoti** [comment]
Closing because the work was already done in #12930

**Comment by bandoti** [comment]
I am reopening this issue as I closed it prematurely—several of the proposed features are not added in #12930, so this should remain open to iterate on that as a baseline.

**Comment by LukeRouleau** [comment]
Yeah, @bandoti, I do think (4) on your user-journey list would be quite useful. Being able to view and explore the high level graph representation of a "model", rather than just the the meta and tensor data in a GGUF, would be very cool. I'm envisioning some equivalent of opening an `.onnx` in netron. Basically, a more seamless version of putting `ggml_graph_dump_dot` in your code

Is that similar to what you meant, or am I off on my own tangent?

**Comment by bandoti** [comment]
@LukeRouleau My initial vision on it is not so much a graph, but more a block-level visual explorer. Initially, a birds-eye view on the data, where double-clicking a block will open a new window with a view on the selected tensor block. These would be color-coded (or greyscale shades) and when moused over on the graphical view, the tree-view of tensors would be highlighted, and vice-versa. Think a sort of disassembly view on the data.

Certainly this new window could expand to provide a graph/tree view of the selected tensors. I am going to (at some point soon) post some drawings of the design I was considering. But a quick description: a column view with three columns (a) list view of meta data; (b) tree view of tensors by grouping; (c) block view of tensor data scaled down but proportional to actual data.

**Comment by ServeurpersoCom** [comment]
I love the idea !!!!

**Comment by ServeurpersoCom** [comment]
I built this as an exercise, now I’ll be able to add it to my llama.cpp dev server page, so when I click on a .gguf, it’ll launch the backend and open the proxied viewer page lol. I’d love some ideas on how to visualize the weights, maybe as heatmaps or graphs; I could even make some in D3.js, like the one on the root page of my domain (showing a real mesh network). Next, I'm planning to experiment with heatmaps to visualize quantization quality and block entropy, I'm already working on a raw visualization using a normalized canvas (-1 to +1); when you drag the view itself, the backend streams the tensor weights as a raw raster, allowing me to plug in any algorithm on the backend side, with the projection streamed in real time so I can navigate large volumes of data smoothly.

https://github.com/ggml-org/llama.cpp/compare/master...ServeurpersoCom:llama.cpp:llama-gguf-viewer

<img width="1707" height="815" alt="Image" src="https://github.com/user-attachments/assets/270f571f-430e-44e6-b97f-814ed1169e87" />
<img width="1707" height="787" alt="Image" src="https://github.com/user-attachments/assets/369c0278-87b6-4a4e-8978-500524a076b0" />
<img width="1707" height="777" alt="Image" src="https://github.com/user-attachments/assets/f9b8681f-64e7-48db-a9aa-2baa8ff3b068" />
<img width="1707" height="777" alt="Image" src="https://github.com/user-attachments/assets/330f9a93-9812-4f08-8faf-3284cd08d701" />

**Comment by ServeurpersoCom** [comment]
Next, I’ll try to visualize the quantization blocks, those little per-group slices (like 32×N tiles), and maybe add some filters to highlight scale or residual patterns. Later, I’ll make it possible to compare two GGUF models differences

<img width="1703" height="817" alt="Image" src="https://github.com/user-attachments/assets/0a209ea9-8f7d-4e7e-93a9-6601ad69ed65" />

**Comment by mtmcp** [comment]
@ServeurpersoCom I am somewhat behind on this, but I will try to get the drawings done for my initial vision, and some ideas for the graph view I would be happy to contribute as soon as I am able.

Regarding the implementation, I would suggest that if you are experimenting on the server side of things, it would probably be beneficial to follow the new SvelteKit #14839 changes to maintain similar Look & Feel as the server along with the other benefits SvelteKit provides.

**Comment by ServeurpersoCom** [comment]
Built this in one day, as KISS as possible: pure C++/GGML + vanilla JS. I still need to transfer FP32 weights in binary instead of JSON to squeeze out more performance.
Next step: time to Sveltify it!

https://www.youtube.com/watch?v=gJSp-Ske-Vc

**Comment by ggerganov** [comment]
@ServeurpersoCom Very cool! Showing the weight values when hovering on the pixels would be useful.

**Comment by ServeurpersoCom** [comment]
> [@ServeurpersoCom](https://github.com/ServeurpersoCom) Very cool! Showing the weight values when hovering on the pixels would be useful.

Sure, on it :)
OBS doesn’t capture Firefox tooltips, so when I tried to show the token hex and info, it didn’t show up on the video 😅

**Comment by ServeurpersoCom** [comment]
I'm noticing some strange artifacts on certain slices of specific models, looks like repeated patterns along one axis, which could either be mathematically expected or a quantization glitch.
When hovering over a pixel, I'll display the raw pre-dequantized value, the exact quantization formula applied, and the resulting FP32 value.
To make visual comparison between different GGUF quantizations easier, I'll upgrade the backend to support multi-threaded tensor streaming. That way, we can open several tabs in browser with different quant levels of the same model and compare them at same slice.

<img width="1705" height="775" alt="Image" src="https://github.com/user-attachments/assets/231137f2-c98a-4aea-a0b4-01730200f64a" />

**Comment by ggerganov** [comment]
Btw can this program stream tensor data directly from HF? If yes, it could become a static web app that would run straight in the browser.

**Comment by ServeurpersoCom** [comment]
For sure we can run the tiny backend on a HF Space. I just need to optimize the communication between the frontend and/or the streaming layer, avoiding resending data that’s already displayed on the client, and possibly using some lightweight image compression technique to speed things up.

**Comment by ServeurpersoCom** [comment]
If we rely only on the GGML public API, the tooltip can safely decode any block using ggml_get_type_traits(type)->to_float, which gives us the FP32 values directly.
That works fine and remains fully future-proof, no need to hardcode every quantization format.

However, if we want to display more detailed info (like scale, zero-point, bias, etc.), I currently have to re-implement each quantization layout manually (q4_0, q5_1, q6_K, ...).
That's fragile, because every time a new format appears in GGML, the viewer will lag behind until I update the parsing logic.

So either we settle for showing only the dequantized FP32 output,
or, if there's a cleaner way to access the raw quantization parameters through GGML without duplicating the internal structures, I'd love to use that instead.
Does GGML expose anything like that, or is to_float really the only safe public entry point?

<img width="785" height="408" alt="Image" src="https://github.com/user-attachments/assets/77e53dc8-d19e-4e1a-96c8-52f76e7548cc" />
<img width="686" height="472" alt="Image" src="https://github.com/user-attachments/assets/864fcb64-f41a-4917-bd8b-8a8918e3eec5" />

**Comment by mtmcp** [comment]
<img width="1575" height="1131" alt="Image" src="https://github.com/user-attachments/assets/5ea8b49f-7e37-4e2f-b312-1f99cae4b22f" />

@ServeurpersoCom, @ggerganov  Here is my initial design ideas. Something not captured in the drawing:

1. Adaptive scaling: when user shrinks window, I would recommend collapsing into a menu in the order: (a) models; (b) meta/tokens; (c) tensors. For example, there would be three widths determined, as the window gets smaller-and-small the less-relevant data gets stuffed into a side menu bar. Never collapsing the heat map though because it is the primary view on the data.
2. Opening multiple models: I was thinking to keep it simple here and just open them in new windows and let the browser do the work there. Just open two windows and view them side-by-side.
3. The tensor tree view as "folders" should remain in sync with the heatmap. So when the tree is selected, the portion on the heatmap becomes focused (using color or a thick border), and vice-versa. 
4. The meta-data would provide a list view on the general meta data when the user toggles that (instead of what's shown the token view).

I'm happy to expand on further ideas just let me know.

**Comment by mtmcp** [comment]
One more thought: I would love to have "deep links" to the active state of the browsing. For example, any time the app's state is updated, to have the URL updated so I can copy/pasted and go back to the exact state. This would be really helpful to share with folks a particular view on the data, or allow us to save/return in a _stateless_ manner. 😊

**Comment by ServeurpersoCom** [comment]
> One more thought: I would love to have "deep links" to the active state of the browsing. For example, any time the app's state is updated, to have the URL updated so I can copy/pasted and go back to the exact state. This would be really helpful to share with folks a particular view on the data, or allow us to save/return in a _stateless_ manner. 😊

You can try it on my dev server (it rebuilds/test often, so no guarantees)

https://www.serveurperso.com/ia/gguf/?grid=1&model=mradermacher/Quasar-2.0-7B-Pro-i1-GGUF/Quasar-2.0-7B-Pro.i1-Q4_K_S.gguf&tensor=blk.0.attn_output.weight&slice=1&min=0.003&max=0.0717&x=335&y=2115#heatmap

I've just implemented full deep-linking support, the viewer state (model, tensor, slice, heatmap settings, etc.) now syncs to the URL, so you can copy, share, or reload any exact view instantly.

I'm still working on a small performance fix to avoid re-reading full tensor slices when min/max values are already known, caching will make it much smoother soon.

**Comment by ServeurpersoCom** [comment]
Same strange thing shows up with unsloth/Magistral-Small-2509-GGUF/* GGUF, and even MistralAI’s own GGUF models, including the BF16 ones, which is pretty weird

<img width="721" height="315" alt="Image" src="https://github.com/user-attachments/assets/850fc1a2-0bf0-4c04-819f-ccb54f6a8fa6" />
<img width="809" height="390" alt="Image" src="https://github.com/user-attachments/assets/031e5a57-97df-4446-bb1c-8714a583ca77" />
<img width="929" height="363" alt="Image" src="https://github.com/user-attachments/assets/28562263-d126-4ec5-928f-c160665eef1b" />

https://www.serveurperso.com/ia/gguf/?grid=0&model=unsloth%2FMagistral-Small-2509-GGUF%2FMagistral-Small-2509-Q6_K.gguf&tensor=token_embd.weight&slice=1&min=-0.0017&max=0.0017&x=0&y=0#heatmap
https://www.serveurperso.com/ia/gguf/?grid=0&model=mistralai%2FMagistral-Small-2509-GGUF%2FMagistral-Small-2509-BF16.gguf&tensor=token_embd.weight&slice=1&min=-0.0017&max=0.0017&x=0&y=0#heatmap

**Comment by mtmcp** [comment]
@ServeurpersoCom Thank you for taking the initiative on this! It already seems to be shaping up nicely.

I cannot comment on the strangeness with these GGUFs. Please do keep us posted if you discover something.

**Comment by ServeurpersoCom** [comment]
Yes, I actually asked MistralAI directly about it here lol : https://www.reddit.com/r/MistralAI/comments/1o0kvpe/weird_repeating_values_in_magistralsmall2509_gguf/

For the viewer, I’d like to finish the Statistics page with a proper frequency histogram scale, and the *essential 2D FFT*, which should reveal some really interesting structures in the frequency spectrum.
Next steps include adding more layer-level plots, splitting the project into multiple files (it’s still one big monolithic blob 😆), fully optimizing the binary + WebSocket communication, and eventually rebuilding the frontend in Svelte for a cleaner architecture.
It’s already fast, bug-free, and 100% KISS just needs some round of cleanup and polish! 🚀

If the issue seen in Magistral ever appears in other GGUFs, it might be worth creating a tiny CLI binary that performs computer-vision analysis on all tensors to automatically detect flat or corrupted regions : could be integrated as a lightweight QA step when publishing new GGUFs !

**Comment by ggerganov** [comment]
I think token embedding tensors can sometimes have zero rows - usually corresponding to padding/placeholder tokens in the vocab.

But the viewer is a bit confusing to me. For example, for this `5120 x 131072` tensor:

<img width="1307" height="176" alt="Image" src="https://github.com/user-attachments/assets/be4bf88a-ab19-4be1-8392-4bc30b2ad109" />

I would expect the width of the display to be 5120 pixels - these we refer to as the columns (i.e. dim 0). And the height should be 131072 - these are the rows (i.e. dim 1). Each row corresponds to the learned embedding of a token in the vocab and we want to see those plotted horizontally since this is the dimension along which we quantize.

In the viewer currently, I think the display is flipped. I.e. the width goes all the way up to 131072 instead of to 5120.

When this rendering issue is resolved, I would expect that we will see some empty rows. I thought that model developers usually put these tokens at the end of the vocab, while here it might seem they are at the beginning, which is a bit strange.

**Comment by ServeurpersoCom** [comment]
That’s exactly the kind of review I’ve been missing : thanks a lot for catching that! 🙏
You’re absolutely right, I’ll flip the visualization so that dim 0 goes horizontal and dim 1 vertical.

Also, in the current (pre-flip) orientation, if you scroll further right, you'll notice some entropy returning in the top rows a few pixels reappearing. This pattern will be more intuitive to analyze once properly oriented.

**Comment by jubruckne** [comment]
Wow this is real great! I wonder if you could hook up an inference loop with eval-callback and step through activation tensors (or kv cache) too? Like a llm-debugger :)

**Comment by ServeurpersoCom** [comment]
@bandoti 
<img width="357" height="735" alt="Image" src="https://github.com/user-attachments/assets/27738191-9119-4bd1-984d-f878006bf4ee" />

Your drawing gave me the idea: https://www.serveurperso.com/ia/gguf/?grid=1&model=mistralai%2FMagistral-Small-2509-GGUF%2FMagistral-Small-2509-BF16.gguf#architecture (During a refactor, don’t rely on the heatmap.)

<img width="504" height="1252" alt="Image" src="https://github.com/user-attachments/assets/1182d763-3182-43d0-9439-effac906fa5e" />

**Comment by ServeurpersoCom** [comment]
> Wow this is real great! I wonder if you could hook up an inference loop with eval-callback and step through activation tensors (or kv cache) too? Like a llm-debugger :)

That would be amazing, but it’ll take quite a few iterations. I’d like to start small and focus on a clean, stable architecture, something useful and easy to evolve. It’s better to aim small and grow it step by step.
To make it meaningful, I'll need FFT2D and fast GPU rendering. I've done work with OpenGL ES compute shaders (or 3D Fresnel RF propagation in a DIY GIS, kind of like a homemade Google Earth) before, so it’s definitely doable.
I’m also still learning more about how Transformers and llama.cpp work, that's actually what keeps me motivated to build these tools :)

**Comment by ServeurpersoCom** [comment]
> I think token embedding tensors can sometimes have zero rows - usually corresponding to padding/placeholder tokens in the vocab.
> 
> But the viewer is a bit confusing to me. For example, for this `5120 x 131072` tensor:
> 
> 
> I would expect the width of the display to be 5120 pixels - these we refer to as the columns (i.e. dim 0). And the height should be 131072 - these are the rows (i.e. dim 1). Each row corresponds to the learned embedding of a token in the vocab and we want to see those plotted horizontally since this is the dimension along which we quantize.
> 
> In the viewer currently, I think the display is flipped. I.e. the width goes all the way up to 131072 instead of to 5120.
> 
> When this rendering issue is resolved, I would expect that we will see some empty rows. I thought that model developers usually put these tokens at the end of the vocab, while here it might seem they are at the beginning, which is a bit strange.

I had a double variable swap that effectively rotated the aspect ratio by 90°, so the orientation looked correct but the scale was wrong: the vertical resolution was compressed by about *25 !

Now that it’s fixed, the hole in token_embd.weight for Magistral is actually much larger than it looked before !

**Comment by ChristianWeyer** [comment]
Did this viewer actually go anywhere @ServeurpersoCom ?

**Comment by ServeurpersoCom** [comment]
> Did this viewer actually go anywhere [@ServeurpersoCom](https://github.com/ServeurpersoCom) ?

It's a project I need to finish! it's remained in the prototype phase due to lack of time

**Comment by ServeurpersoCom** [comment]
I really want to develop this GGUF visualizer, and since I finished the backend of a specialized GGML AI inference from scratch and I'm at the embedded Vite/TS/Svelte interface part and it's really practical, I'll be able to resume coding in a much cleaner way than with my VanillaJS proto with the experience!

**Comment by lutzroeder** [comment]
[Netron](https://github.com/lutzroeder/netron) has experimental support for GGUF graphs.

<img width="1136" height="880" alt="Image" src="https://github.com/user-attachments/assets/deda24a8-1542-42d0-bb70-77d3f6a43f55" />

**Comment by ChristianWeyer** [comment]
> [Netron](https://github.com/lutzroeder/netron) has experimental support for GGUF graphs.

Ohhhh, hey Lutz - servus! :-D

**Comment by ngxson** [comment]
closing in this is already implemented
